vnext/Microsoft.ReactNative/CoalescingEventEmitter.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CoalescingEventEmitter.h"
+#include "CoalescingEventEmitter.g.cpp"
+#include "DynamicWriter.h"
+
+namespace winrt::Microsoft::ReactNative::implementation {
+
+/*static*/ ReactPropertyId<CoalescingEventEmitter> CoalescingEventEmitter::Property() noexcept {
+  static ReactPropertyId<CoalescingEventEmitter> propId{L"ReactNative.CoalescingEventEmitter", L"Emitter"};
+  return propId;
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+/*static*/ winrt::Microsoft::ReactNative::CoalescingEventEmitter CoalescingEventEmitter::FromContext(
+    Mso::React::IReactContext const &context) noexcept {
+  return context.Properties()
+      .Get(CoalescingEventEmitter::Property().Handle())
+      .as<winrt::Microsoft::ReactNative::CoalescingEventEmitter>();
+}
+
+CoalescingEventEmitter::CoalescingEventEmitter(Mso::CntPtr<Mso::React::IReactContext> &&context) noexcept
+    : m_context(std::move(context)) {}
+
+void CoalescingEventEmitter::EmitJSEvent(int64_t coalesceKey, JSValueArgWriter const &paramsArgWriter) noexcept {
+  std::unique_lock lock(m_mutex);
+
+  for (auto &evt : m_eventQueue) {
+    if (evt.coalesceKey && *evt.coalesceKey == coalesceKey) {
+      evt.expired = true;
+    }
+  }
+
+  auto paramsWriter = winrt::make_self<DynamicWriter>();
+  paramsArgWriter(*paramsWriter);
+
+  CoalescingEvent newEvent;
+  newEvent.coalesceKey = coalesceKey;
+  newEvent.args = paramsWriter->TakeValue();
+
+  m_eventQueue.push_back(std::move(newEvent));
+  EnsureRenderCallbackRegistered();
+}
+
+void CoalescingEventEmitter::EmitDurableJSEvent(JSValueArgWriter const &paramsArgWriter) noexcept {
+  std::unique_lock lock(m_mutex);
```

Comment:
We remove events on the JS thread, so we still need to lock. It would be nice to do something InvokeElsePost style to the UI thread we associate with the context.

Reply:
We do need to access the queue on the JS thread though, since we need to throttle to how often the JS thread is able to process (which may be less than UI thread)